flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/MultinomialNaiveBayes.scala

```diff
@@ -0,0 +1,900 @@
+    override val defaultValue: Option[Int] = Some(0)
+  }
+
+  case object P3 extends Parameter[Int] {
+    override val defaultValue: Option[Int] = Some(0)
+  }
+
+  case object SR1 extends Parameter[Int] {
+    override val defaultValue: Option[Int] = Some(0)
+  }
+
+  case object S1 extends Parameter[Int] {
+    override val defaultValue: Option[Int] = Some(0)
+  }
+
+  case object R1 extends Parameter[Int] {
+    override val defaultValue: Option[Int] = Some(0)
+  }
+
+  // ======================================== Factory Methods ======================================
+  def apply(): MultinomialNaiveBayes = {
+    new MultinomialNaiveBayes()
+  }
+
+  // ====================================== Operations =============================================
+  /**
+   * Trains the models to fit the training data. The resulting
+   * [[MultinomialNaiveBayes.wordRelatedModelData]] and
+   * [[MultinomialNaiveBayes.classRelatedModelData]] are stored in the [[MultinomialNaiveBayes]]
+   * instance.
+   */
+
+  implicit val fitNNB = new FitOperation[MultinomialNaiveBayes, (String, String)] {
+    /**
+     * The [[FitOperation]] used to create the model. Requires an instance of
+     * [[MultinomialNaiveBayes]], a [[ParameterMap]] and the input data set. This data set
+     * maps (string -> string) containing (label -> text, words separated by ",")
+     * @param instance of [[MultinomialNaiveBayes]]
+     * @param fitParameters, additional parameters
+     * @param input, the to processed data set
+     */
+    override def fit(instance: MultinomialNaiveBayes,
+                     fitParameters: ParameterMap,
+                     input: DataSet[(String, String)]): Unit = {
+
+      val resultingParameters = instance.parameters ++ fitParameters
+
+      //Count the amount of documents for each class.
+      // 1. Map: replace the document text by a 1
+      // 2. Group-Reduce: sum the 1s by class
+      val documentsPerClass: DataSet[(String, Int)] = input.map { input => (input._1, 1)}
+        .groupBy(0).sum(1) // (class name -> count of documents)
+
+      //Count the amount of occurrences of each word for each class.
+      // 1. FlatMap: split the document into its words and add a 1 to each tuple
+      // 2. Group-Reduce: sum the 1s by class, word
+      var singleWordsInClass: DataSet[(String, String, Int)] = input
+        .flatMap(new SingleWordSplitter())
+        .groupBy(0, 1).sum(2) // (class name -> word -> count of that word)
+
+      //POSSIBILITY 2: all words in class (order of operators)
+      //SCHNEIDER/RENNIE 1: ignore/reduce word frequency information
+        //the allWordsInClass data set does only contain distinct
+        //words for schneiders approach: ndw(cj), nothing changes for rennies approach
+
+      val p2 = resultingParameters(P2)
+
+      val sr1 = resultingParameters(SR1)
+
+      var allWordsInClass: DataSet[(String, Int)] =
+        null // (class name -> count of all words in that class)
```

Comment:
As above, perhaps move `null` to the previous line. [I'm not sure about this. Maybe this is just a matter of personal taste. Perhaps someone else will weigh in at some point.]